src/WebJobs.Extensions/Extensions/Timers/Listener/TimerListener.cs

```diff
@@ -79,41 +80,39 @@ internal System.Timers.Timer Timer
             {
                 // check to see if we've missed an occurrence since we last started.
                 // If we have, invoke it immediately.
-                _scheduleStatus = await ScheduleMonitor.GetStatusAsync(_timerName);
-                TimeSpan pastDueDuration = await ScheduleMonitor.CheckPastDueAsync(_timerName, now, _schedule, _scheduleStatus);
+                ScheduleStatus = await ScheduleMonitor.GetStatusAsync(_timerName);
+                TimeSpan pastDueDuration = await ScheduleMonitor.CheckPastDueAsync(_timerName, now, _schedule, ScheduleStatus);
                 isPastDue = pastDueDuration != TimeSpan.Zero;
+            }
 
-                if (_scheduleStatus == null)
+            if (ScheduleStatus == null)
+            {
+                // no schedule status has been stored yet, so initialize
+                ScheduleStatus = new ScheduleStatus
                 {
-                    // no schedule status has been stored yet, so initialize
-                    _scheduleStatus = new ScheduleStatus
-                    {
-                        Last = default(DateTime),
-                        Next = _schedule.GetNextOccurrence(now)
-                    };
-                }
+                    Last = default(DateTime),
+                    Next = _schedule.GetNextOccurrence(now)
+                };
             }
 
             if (isPastDue)
             {
                 _trace.Verbose(string.Format("Function '{0}' is past due on startup. Executing now.", _timerName));
-                await InvokeJobFunction(now, true);
+                await InvokeJobFunction(now, isPastDue: true);
             }
             else if (_attribute.RunOnStartup)
             {
                 // The job is configured to run immediately on startup
                 _trace.Verbose(string.Format("Function '{0}' is configured to run on startup. Executing now.", _timerName));
-                await InvokeJobFunction(now);
+                await InvokeJobFunction(now, runOnStartup: true);
             }
 
             // log the next several occurrences to console for visibility
             string nextOccurrences = TimerInfo.FormatNextOccurrences(_schedule, 5);
             _trace.Verbose(nextOccurrences);
 
-            // start the timer
-            now = DateTime.Now;
-            DateTime nextOccurrence = _schedule.GetNextOccurrence(now);
-            TimeSpan nextInterval = nextOccurrence - now;
+            // start the timer            
```

Comment:
Might as well add some detailed comments here, that we need to start the timer AFTER the function invocation (the comments you made below)